### PR TITLE
Updating Python scripting chapter with Zipf's Law narrative

### PR DESCRIPTION
**Pull request by DamienIrving**

### CODE REVIEW

File: py-rse/py-scripting.Rmd [hunk 1]
```diff
@@ -135,7 +133,262 @@ optional arguments:
...
+
+    return word_counts
+
+
+def main(args):
+    """Run the command line program."""
+    with args.infile as reader:
+        word_counts = count_words(reader)
+    word_counts = sort_counts(word_counts, args.sortby)
+    report_results(sys.stdout, word_counts)
```
[the remark this answers is not in the document]
Comment (DamienIrving) [reply py-rse/py-scripting.Rmd]: Good suggestion!

File: py-rse/py-scripting.Rmd [hunk 2]
```diff
@@ -135,7 +133,262 @@ optional arguments:
...
+import argparse
+from collections import Counter
+
+
+def report_results(writer, results):
+    """Report results to stream."""
+    
+    writer = csv.writer(writer)
+    for (key, value) in results:
+        writer.writerow((key, value))
```
[the remark this answers is not in the document]
Comment (DamienIrving) [reply py-rse/py-scripting.Rmd]: Thanks. That's much neater.

**Comment by DamienIrving** [comment]
This is now ready for review.

Based on feedback from others, I've now developed all three scripts (`countwords.py`, `collate.py` and `plotcounts.py`) in this Python scripting lesson. At the conclusion of the lesson the first two scripts are in their final form, whereas `plotcounts.py` will have some improvements made to it in subsequent chapters, as outlined in the updated [narrative](https://github.com/DamienIrving/merely-useful.github.io/blob/cmdline-programs/py-rse/narrative.md).

@cwickham - I hope this helps with #306? Instead of having to develop the entire `collate.py` script from scratch, the new [narrative](https://github.com/DamienIrving/merely-useful.github.io/blob/cmdline-programs/py-rse/narrative.md) suggests that for the git lesson you just need to make a minor update to `plotcounts.py` so that it plots a log-log graph (it's literally only 1 or 2 lines of code that need to be changed).

@elliewix - The new [narrative](https://github.com/DamienIrving/merely-useful.github.io/blob/cmdline-programs/py-rse/narrative.md) also hopefully makes the advanced git lesson update a little easier.

File: py-rse/py-scripting.Rmd [hunk 3]
```diff
@@ -135,15 +129,399 @@ optional arguments:
...
+
+Using `script_template.py` as a guide,
+take this plotting code and write a new python script called `plotcounts.py`.
+The script should:
+
+a. Use the `type=argparse.FileType('r')`, `nargs='?'` and `default='-'` options
+   for the input file argument (i.e. similar to the `countwords.py` script)
+   so that `plotcounts.py` uses standard input if no csv file is given.
+
+b. Add an optional `--xlim` option so that the user can change the x-axis bounds.
```
Comment (cwickham) [inline py-rse/py-scripting.Rmd]: It doesn't seem to have come up in this chapter but why two dashes here?  

The "Key points" talk about using short and long command line options, but that doesn't seem to be explicitly talked about in the chapter.  Might also need to comment on when to use positional vs. optional arguments.  Does `argparse` do anything different based on the name you give an argument, i.e. if it has no, one or two dashes?

Comment (cwickham) [reply py-rse/py-scripting.Rmd]: Another thought that occurred as I started on command line git: do you ever talk about where one might be saving these python programs? Do their relative locations matter?

Comment (DamienIrving) [reply py-rse/py-scripting.Rmd]: @SaraMati had the same question in relation to the shell lessons. I think we're ultimately going to have to have a zip file that people download and unzip on their Desktop with all the data files and various directories we use in it. In the meantime, I think we can just assume they are saving the scripts in the `/zipfs_law/bin/` directory.

File: py-rse/py-scripting.Rmd [hunk 4]
```diff
@@ -68,24 +68,18 @@ The simplest way to explain how `argparse` works is by example,
 so let's create a short Python script called `script_template.py`:
```
[the remark this answers is not in the document]
Comment (DamienIrving) [reply py-rse/py-scripting.Rmd]: Good point. I've added a detailed description of what's happening inside `script_template.py`.

File: py-rse/py-scripting.Rmd [hunk 5]
```diff
@@ -135,15 +129,399 @@ optional arguments:
...
+Let's call the script `countwords.py`:
+
+```python
+"""Count the occurences of all words in a text and write them to a CSV-file."""
+import sys
+import re
+import argparse
+from collections import Counter
+
+def report_results(results, top_n=None):
```
[the remark this answers is not in the document]
Comment (DamienIrving) [reply py-rse/py-scripting.Rmd]: Excellent points. I've updated things by replacing `report_results` with a `collection_to_csv` function that is much better described in text.

File: py-rse/py-scripting.Rmd [hunk 6]
```diff
@@ -135,15 +129,399 @@ optional arguments:
...
+    args = parser.parse_args()
+    main(args)
+```
+
+The [argparse][argparse] library comes with many different options
+for defining the characteristics of command line arguments.
+For the `infile` option defined above,
+the number of expected arguments (`nargs`) is set to `?`.
+This means that one argument (i.e. a file name)
+will be consumed from the command line if possible.
```
[the remark this answers is not in the document]
Comment (DamienIrving) [reply py-rse/py-scripting.Rmd]: Not sure how to make it clearer, so I've left this as is.

File: py-rse/py-scripting.Rmd (hunk 3 above)
Comment (DamienIrving) [reply py-rse/py-scripting.Rmd]: I've removed the reference to long and short options from the key points and I've created a call out box to talk about the different between positional and optional arguments (and short and long optional arguments).

**Comment by DamienIrving** [comment]
Thanks for the review, @cwickham! I've made the changes you suggested - it should now be ready for merging.

**Comment by DamienIrving** [comment]
@lwjohnst86 @cwickham Should I go ahead and merge? The updated narrative file in this PR will be useful for @SaraMati in doing the advanced git lesson, so it would be good to get this one merged.

**Comment by lwjohnst86** [comment]
Yes, I'll merge. Still need to fix the build error. I'll work on it tomorrow and Tuesday.

**Pull request merged**